### PR TITLE
add: minesweeper PDA game

### PR DESCRIPTION
**Pull request by ROdenFL**
## What Does This PR Do

Adds new basic pda app "Games" with only one game: Minesweeper.
RMB for flags, LMB for check cell.

Also, i'd like to do less string operations, but i don't see any ways of it for now.

CSS and JS modificated by @AyIong

## Why It's Good For The Game

People can play minesweeper, instead of doing work, just like irl.

## Images of changes

image
image
image

## Testing

Played a lot. Winned sometimes.

<hr>

### Declaration

- [x] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/docs/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
  
  
  <hr>

## Changelog

:cl:
add: minesweeper game in pda
/:cl:

### CODE REVIEW

File: code/modules/pda/games.dm [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+	name = "base game"
+	icon = "gamepad"
+	hidden = TRUE
+
+
+/datum/data/pda/app/game/proc/get_game_info()
+	return list(name = name, icon = icon, id = UID())
+
+
+// Minesweeper
```
Comment (dj-34) [inline code/modules/pda/games.dm:37]: Probably need a separate file for the games. Putting all new (future) games in one file is hell.

File: code/modules/pda/games.dm [hunk 2]
```diff
@@ -0,0 +1,284 @@
+// Games Hub
+
+/datum/data/pda/app/games
+	name = "Games"
+	icon = "gamepad"
+	template = "pda_games"
+
```
Comment (dj-34) [inline code/modules/pda/games.dm:7]:
```suggestion
```

Comment (dj-34) [reply code/modules/pda/games.dm:7]: So many double entries. I don't know what they're for.

File: tgui/packages/tgui/styles/main.scss [hunk 3]
```diff
@@ -61,6 +61,7 @@
 @include meta.load-css('./interfaces/OreRedemption.scss');
 @include meta.load-css('./interfaces/PanDEMIC.scss');
 @include meta.load-css('./interfaces/PDA.scss');
+@include meta.load-css('./interfaces/pda_minesweeper.scss');
```
Comment (dj-34) [inline tgui/packages/tgui/styles/main.scss:64]: Also maybe should be separated somehow, maybe in code position

**Comment by Crazyhair213** [comment]
We’re getting mini games!!! Dchat mafia coming to a theater near you

File: tgui/packages/tgui/styles/main.scss (hunk 3 above)
Comment (ROdenFL) [reply tgui/packages/tgui/styles/main.scss:64]: don`t see any need in this

File: code/modules/pda/games/minesweeper.dm [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+						if(minesweeper_matrix[params["X"]][params["Y"]]["bomb"])
+							flagged_bombs -= 1
+					else
+						minesweeper_matrix[params["X"]][params["Y"]]["flag"] = TRUE
+						setted_flags += 1
+						if(minesweeper_matrix[params["X"]][params["Y"]]["bomb"])
+							flagged_bombs += 1
+
+			check_win(ui.user)
+			SStgui.update_uis(pda)
```
Comment (AyIong) [inline code/modules/pda/games/minesweeper.dm:79]:
```suggestion
/datum/data/pda/app/game/minesweeper/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
	if(..())
		return

	if(ignore_touches)
		return

	switch(action)
		if("Square")
			switch(params["mode"])
				if("bomb")
					if(first_touch)
						generate_matrix(params["X"], params["Y"])
					open_cell(params["X"], params["Y"])

					if(minesweeper_matrix[params["X"]][params["Y"]]["bomb"])
						on_loose(ui.user)
						return

				if("flag")
					if(first_touch || minesweeper_matrix[params["X"]][params["Y"]]["open"])
						return

					if(minesweeper_matrix[params["X"]][params["Y"]]["flag"])
						minesweeper_matrix[params["X"]][params["Y"]]["flag"] = FALSE
						setted_flags -= 1
						if(minesweeper_matrix[params["X"]][params["Y"]]["bomb"])
							flagged_bombs -= 1
					else
						minesweeper_matrix[params["X"]][params["Y"]]["flag"] = TRUE
						setted_flags += 1
						if(minesweeper_matrix[params["X"]][params["Y"]]["bomb"])
							flagged_bombs += 1

			check_win(ui.user)
```

ui_act updates UI itself

File: code/modules/pda/games/minesweeper.dm [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+		playsound(get_turf(pda), 'sound/machines/ping.ogg', 20, 1)
+	addtimer(CALLBACK(src, PROC_REF(make_empty_matr)), 5 SECONDS)
+	add_into_leaders(user, world.time - start_time)
+
+/datum/data/pda/app/game/minesweeper/proc/add_into_leaders(mob/user, game_time)
+	var/nickname = tgui_input_text(user, "You finished the game into [game_time/10] seconds.\n Write a nickname to save your result in leaderboard.\n", "Minesweeper", "", 10)
+	if(!nickname)
+		return
+	else
+		leaderboard += list(list("name" = nickname, "time" = "[game_time/10]"))
```
Comment (AyIong) [inline code/modules/pda/games/minesweeper.dm:99]:
```suggestion
/datum/data/pda/app/game/minesweeper/proc/add_into_leaders(mob/user, game_time)
	var/nickname = tgui_input_text(user, "You finished the game into [game_time/10] seconds.\n Write a nickname to save your result in leaderboard.\n", "Minesweeper", "", 10)
	if(isnull(nickname))
		return

	leaderboard += list(list("name" = nickname, "time" = "[game_time/10]"))
```

Use `isnull()` for text_input

File: tgui/packages/tgui/interfaces/pda/pda_games.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import { useBackend } from '../../backend';
+import { Box, Button, ImageButton, Section } from '../../components';
+import { Icon, IconStack } from '../../components/Icon';
+
+export const pda_games = (props, context) => {
+  const { act, data } = useBackend(context);
+
+  const { games } = data;
```
Comment (AyIong) [inline tgui/packages/tgui/interfaces/pda/pda_games.js:8]:
```suggestion
  const { act, data } = useBackend(context);
  const { games } = data;
```

File: tgui/packages/tgui/interfaces/pda/pda_minesweeper.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
+import { Fragment } from 'inferno';
+import { useBackend, useLocalState } from '../../backend';
```
Comment (AyIong) [inline tgui/packages/tgui/interfaces/pda/pda_minesweeper.js:2]:
```suggestion
import { useBackend, useLocalState } from '../../backend';
```

**Review by AyIong — COMMENTED** [review]
Some things

File: code/modules/pda/games/minesweeper.dm (hunk 5 above)
Comment (ROdenFL) [reply code/modules/pda/games/minesweeper.dm:99]: i don`t want to see empty input too

**Comment by Danchi299** [comment]
> Also, i`d like to do less string operations, but i don`t see any ways of it for now.

`Backticks` have special interactions in markdown, would suggest 'quote marks' instead

File: tgui/packages/tgui/styles/main.scss (hunk 3 above)
Comment (Danchi299) [reply tgui/packages/tgui/styles/main.scss:64]: if we are getting entire PDA sub category for Games, then yeah, I think putting it in something like 
`./interfaces/pda-games/minesweeper.scss` would be nice

**Comment by ROdenFL** [comment]
> > Also, i`d like to do less string operations, but i don`t see any ways of it for now.
> 
> `Backticks` have special interactions in markdown, would suggest 'quote marks' instead

it would be great if you show an example, or describe in more words. Im not native english speaker and low-experienced coder. there is nothing i understood)

**Comment by ROdenFL** [comment]
> > > Also, i`d like to do less string operations, but i don`t see any ways of it for now.
> > 
> > 
> > `Backticks` have special interactions in markdown, would suggest 'quote marks' instead
> 
> it would be great if you show an example, or describe in more words. Im not native english speaker and low-experienced coder. there is nothing i understood)

fuck, i understood

**Comment by BakKeener** [comment]
I want this very badly, would also encourage little minigames from other people and maybe have nano-mon make a come back by adding little rooms for you to play against other players with.

**Comment by Denthamos** [comment]
Would it be possible to make the mines themselves look like syndiebombs in the future?  Would be fitting/amusing if possible.

**Comment by ROdenFL** [comment]
> Would it be possible to make the mines themselves look like syndiebombs in the future?  Would be fitting/amusing if possible.

I use Font Awesome for bombs. I think, it possible to replace it with image buttons, but it will be quite hard work

**Comment by AyIong** [comment]
> > Would it be possible to make the mines themselves look like syndiebombs in the future?  Would be fitting/amusing if possible.
> 
> I use Font Awesome for bombs. I think, it possible to replace it with image buttons, but it will be quite hard work

You can just put DmIcon instead Icon

**Comment by ROdenFL** [comment]
if anybody could make sprite for it, i may try to use it here then

**Comment by ROdenFL** [comment]
> JS looks pretty good, but the naming for the interface files and stylesheets should be changed to match the naming convention of existing interfaces. 

Interfaces have similar names as other pda interfaces. Stylesheet file copys interface name, but with scss extension

**Comment by AyIong** [comment]
> This pull request seems to be stale as there have been no changes in 14 days, please make changes within 7 days or the PR will be closed. If you believe this is a mistake, please inform a development team member on Discord.

_Hush_

**Comment by ROdenFL** [comment]
> This pull request seems to be stale as there have been no changes in 14 days, please make changes within 7 days or the PR will be closed. If you believe this is a mistake, please inform a development team member on Discord.

*Bonk*

File: code/modules/pda/games/minesweeper.dm [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+	leaderboard += list(list("name" = nickname, "time" = "[game_time/10]"))
+
+/datum/data/pda/app/game/minesweeper/proc/on_loose(mob/user)
+	ignore_touches = TRUE
+	if(!pda.silent)
+		playsound(get_turf(pda), 'sound/effects/explosionfar.ogg', 50, 1)
+	addtimer(CALLBACK(src, PROC_REF(make_empty_matr)), 3 SECONDS)
+
+/datum/data/pda/app/game/minesweeper/proc/make_empty_matr(pay = TRUE)
```
Comment (Henri215) [inline code/modules/pda/games/minesweeper.dm:107]: pay?
And why not matrix instead of matr 😢

File: code/modules/pda/games/minesweeper.dm [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	setted_flags = 0
+	opened_cells = 0
+	var/list/possible_list = list()
+	var/num_x = text2num(x)
+	var/num_y = text2num(y)
+	var/count = 0
+
+	for(var/i in 1 to generation_rows)
+		for(var/j in 1 to generation_columns)
+			if((i in list(num_x-1, num_x, num_x+1)) && (j in list(num_y-1, num_y, num_y+1)))
```
Comment (Henri215) [inline code/modules/pda/games/minesweeper.dm:129]:
```suggestion
			if((i in list(num_x - 1, num_x, num_x + 1)) && (j in list(num_y - 1, num_y, num_y + 1)))
```

File: code/modules/pda/games/minesweeper.dm [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+/datum/data/pda/app/game/minesweeper/proc/check_win(mob/user)
+	if(flagged_bombs == generation_bombs && \
+		setted_flags == generation_bombs && \
+		opened_cells == (generation_rows * generation_columns - generation_bombs))
+		on_win(user)
+
+/datum/data/pda/app/game/minesweeper/proc/on_win(mob/user)
+	ignore_touches = TRUE
+	if(!pda.silent)
+		playsound(get_turf(pda), 'sound/machines/ping.ogg', 20, 1)
```
Comment (Henri215) [inline code/modules/pda/games/minesweeper.dm:90]:
```suggestion
		playsound(get_turf(pda), 'sound/machines/ping.ogg', 20, TRUE)
```

File: code/modules/pda/games/minesweeper.dm [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+	addtimer(CALLBACK(src, PROC_REF(make_empty_matr)), 5 SECONDS)
+	add_into_leaders(user, world.time - start_time)
+
+/datum/data/pda/app/game/minesweeper/proc/add_into_leaders(mob/user, game_time)
+	var/nickname = tgui_input_text(user, "You finished the game into [game_time/10] seconds.\n Write a nickname to save your result in leaderboard.\n", "Minesweeper", "", 10)
```
Comment (Henri215) [inline code/modules/pda/games/minesweeper.dm:95]:
```suggestion
	var/nickname = tgui_input_text(user, "You finished the game in [game_time / 10] seconds.\n Write a nickname to save your result on the leaderboard.\n", "Minesweeper", "", 10)
```

File: code/modules/pda/games/minesweeper.dm [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+	if(!nickname)
+		return
+
+	leaderboard += list(list("name" = nickname, "time" = "[game_time/10]"))
+
+/datum/data/pda/app/game/minesweeper/proc/on_loose(mob/user)
+	ignore_touches = TRUE
+	if(!pda.silent)
+		playsound(get_turf(pda), 'sound/effects/explosionfar.ogg', 50, 1)
```
Comment (Henri215) [inline code/modules/pda/games/minesweeper.dm:104]:
```suggestion
		playsound(get_turf(pda), 'sound/effects/explosionfar.ogg', 50, TRUE)
```

**Review by Henri215 — COMMENTED** [review]
Specifiy on the changelog where the game is available.

**Comment by Henri215** [comment]
Tested everything except the leaderboard, i dont know how to win this game, it was looking good.

**Comment by 1080pCat** [comment]
cant wait to learn how to speed run this..

**Comment by SteelSlayer** [comment]
Author has linked their account.

**Pull request merged**